File: main_probability.cpp
#include <vector>
#include <iostream>
#include <cstdlib>
#include <ctime>

using std::cout;
using std::endl;
using std::vector;

int main()
{
	srand(time(NULL));

	/*! 13 card for each suit.
	 ** Ace is 11, J,Q and K are 10 */
	vector<int> deck = {
		2,3,4,5,6,7,8,9,10,10,10,10,11,
		2,3,4,5,6,7,8,9,10,10,10,10,11,
		2,3,4,5,6,7,8,9,10,10,10,10,11,
		2,3,4,5,6,7,8,9,10,10,10,10,11
	};

	/*! Take one random integer from deck container */
	auto randomCard = [&]() -> int {
		return deck[rand() % deck.size()];
	};

	/*! Return true if dealer has [18;21] sum, otherwise false */
	/** sum - the first (faced up) dealer card */
	auto dealerWins = [&](int sum = 0) -> bool {
		while(sum <= 17) {
			int card = randomCard();
			/** dealer take Ace card as 1 (not as 11)
			 ** but only if otherwise he fails */
			if(card == 11 && sum + card > 21)
				card = 1;
			sum += card;
			if(sum > 21)
				return false;
		} 
		return true;
	};

	/*! Return the probability of the dealer to win when
	 ** he starts with "faced_up" card */
	/** accurance - the quantity of experiments */
	auto dealerStartsWith = [&](int faced_up, int accurance = 1000) -> double {
		int wins = 0;
		int fails = 0;
		for(int i = 0; i < accurance; i++) {
			(dealerWins(faced_up))? wins++ : fails++;
		}
		return (double) wins / (wins + fails) * 100;
	};

	for(int i = 2; i <= 11; i++)
		cout << "Faced up card is " << i << " : " << dealerStartsWith(i) << '%' << endl;
}

File: main_probability.clj
(ns BlackJack)

;; !Support part

(def row '(2 3 4 5 6 7 8 9 10 10 10 10 11))
(def deck (list row row row row))

(defn rcard []
  "Get random card from a deck"
  (rand-nth (rand-nth deck)))

;; !Useful part

(defn dealer-plays
  "Simulate dealer's game when he hits another card"
  [& {:keys [sum]}]
  (cond (> sum 21) false ;; it is fail
        (> sum 17) true  ;; deals doesn't have to hit
        :else (dealer-plays :sum (+ sum
                                  (let [r (rcard)]  ;; ace cand be both 11 and 1
                                    (if (and (= r 11) (> sum 10)) 1 r))))))

(defn dealer-starts-with
  "Simulate dealer's game for [acc] times"
  [& {:keys [faced-up wins fails acc]}]
  (if (> acc 1)
    (if (dealer-plays :sum faced-up)
      (dealer-starts-with :faced-up faced-up, :wins (+ wins 1), :fails fails,       :acc (- acc 1))  ;; dealer wins
      (dealer-starts-with :faced-up faced-up, :wins wins,       :fails (+ fails 1), :acc (- acc 1))) ;; dealer fails
    (float (/ wins (+ wins fails)))))

;; !I/O part

(defn print-statistics
  "Final result has to be like 'Faced up with 4 : 50%'"
  [& {:keys [id value] :or {id -1 value -1}}]
  (println (str "Faced up with " id " : " value)))

(defn give-statistics
  "Gather information from dealer-starts-with for i'th card"
  [i]
  (map-to-vec (assoc {} :id i :value (* 100 (dealer-starts-with :faced-up i, :wins 0, :fails 0, :acc 1000)))))

(give-statistics 2)
;; !Helpers' part

(defn wide-into
  ([a] a)
  ([a b] (into a b))
  ([a b & args]
    (if-not (next args)
      (into a (apply into b args))
      (recur (into a b) (first args) (rest args)))))

(defn map-to-vec
  "Converts map to vector"
  [arg]
  (apply into `(~@arg))) ;; using Clojure
  ;;(apply wide-into (map identity arg))) ;; using wide-into

;; !Main part

(defn main
  []
  (->>
   (range)
   (take 12)
   (drop 2)
   (map give-statistics)
   (map #(apply print-statistics %))))
(main)
